src/skozik/krestikinoliki/intelligence/Intelligence2D.java

```diff
@@ -0,0 +1,187 @@
+            if (victPerTurn == null) {
+                victPerTurn = new HashMap<>();
+            }
+            Integer wins = victPerTurn.get(forecastedLevel);
+            if (wins == null) {
+                wins = 1;
+            } else {
+                wins++;
+            }
+            victPerTurn.put(forecastedLevel, wins);
+            victories.put(symbolFriend, victPerTurn);
+            return;
+        }
+        int[][] emptyCoordinates = thisTestTurn.getEmptyCoordinates();
+        if (emptyCoordinates.length <= 1) {
+            return;
+        }
+        for (int i = 0; i < emptyCoordinates.length; i++) {
+            calculateVictoryValues(
+                thisTestTurn,
+                emptyCoordinates[i],
+                symbolEnemy,
+                symbolFriend,
+                victories,
+                forecastedLevel + 1);
+        }
+    }
+
+    @Override
+    public int[] forecastNextStep(IField field, IGamer friend, IGamer enemy) {
+        char symbolFriend = friend.getSymbol();
+        char symbolEnemy = enemy.getSymbol();
+        int[] fieldSize = field.getSize();
+
+        VictoriesTurns[][] victPerTurnFriend = new VictoriesTurns[fieldSize[0]][fieldSize[1]];
+        VictoriesTurns[][] victPerTurnEnemy = new VictoriesTurns[fieldSize[0]][fieldSize[1]];
+        for (int i = 0; i < fieldSize[0]; i++) {
+            for (int j = 0; j < fieldSize[1]; j++) {
+                int[] coordinates = new int[]{i, j};
+                Map<Character, Map<Integer, Integer>> victoriesFriend = new HashMap<>();
+                calculateVictoryValues(field, coordinates, symbolFriend, symbolEnemy, victoriesFriend, 0);
+                Map<Character, Map<Integer, Integer>> victoriesEnemy = new HashMap<>();
+                calculateVictoryValues(field, coordinates, symbolEnemy, symbolFriend, victoriesEnemy, 0);
+                victPerTurnFriend[i][j] = calculateVictories(victoriesFriend, symbolFriend, i, j);
+                victPerTurnEnemy[i][j] = calculateVictories(victoriesEnemy, symbolEnemy, i, j);
+            }
+        }
+        VictoriesTurns enemyTurn = chooseClosestWin(field, fieldSize[0], fieldSize[1], victPerTurnEnemy);
+        VictoriesTurns friendTurn = chooseClosestWin(field, fieldSize[0], fieldSize[1], victPerTurnFriend);
+        if (friendTurn.turn == 0) {
+            return new int[]{friendTurn.x, friendTurn.y};
+        }
+        if (enemyTurn.turn == 0) {
+            return new int[]{enemyTurn.x, enemyTurn.y};
+        }
+        if ((friendTurn.turn <= enemyTurn.turn)
+            || (victPerTurnFriend[friendTurn.x][friendTurn.y].victories > victPerTurnFriend[enemyTurn.x][enemyTurn.y].victories)) {
+            return new int[]{friendTurn.x, friendTurn.y};
+        }
+        return new int[]{enemyTurn.x, enemyTurn.y};
+    }
+
+    private VictoriesTurns chooseClosestWin(IField field, int sizeX, int sizeY, VictoriesTurns[][] victoriesTurns) {
```

Comment:
думаю стоит добавить javadoc

Reply:
так же может вынесем в отдельные методы? думаю так будет удобнее 3 метода примерно я вижу. Думаю логику не портит

Reply:
javadoc added. I have not elaborate solution with splittiing this method in separate methods because this needs methods which return 2 int values simultaneously and leads to solution with complicated data structures.